### PR TITLE
Restricciones hotel image test y seeders

### PR DESCRIPTION
**Pull request by jorgomde**
### **PR Type**
Enhancement, Tests, Other

___

### **Description**
- Added comprehensive image management for hotels, including upload, update, delete, and retrieval functionalities.

- Introduced new tests for hotel image operations to ensure functionality and reliability.

- Implemented seeding for hotel images with a new management command.

- Updated hotel services and views to handle image-related operations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>services.py</strong><dd><code>Add hotel image management methods and validations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/services.py

<li>Added methods for hotel image management (upload, update, delete, <br>retrieve).<br> <li> Introduced validation logic for image operations (e.g., cover image <br>restrictions, max image count).<br> <li> Enhanced serialization for hotel images.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/178/files#diff-095282d677bd0ecb4cfb20d05ac56a005b0510c195f4938da3adc3aa1e7b53f7">+78/-9</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>views.py</strong><dd><code>Add endpoints for hotel image operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/views.py

<li>Added endpoints for hotel image operations (upload, update, delete, <br>retrieve).<br> <li> Integrated service layer methods for image handling.<br> <li> Enhanced response handling for image-related actions.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/178/files#diff-29eae86551ee9fe269942e1ca9a284e2c7c05aa7b4f183bd215252e670bce1fd">+82/-8</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_views.py</strong><dd><code>Add tests for hotel image operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/tests/test_views.py

<li>Added tests for hotel image operations (upload, update, delete, <br>retrieve).<br> <li> Introduced utility for creating test images.<br> <li> Verified edge cases like invalid image uploads.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/178/files#diff-90f49e9fda72cb5d99aad165e5770e0eaf1b942929b472eb40c76ec7cb6f5e60">+135/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Other</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>seed.py</strong><dd><code>Update seed command to include hotel images</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/management/commands/seed.py

- Added call to seed hotel images in the main seed command.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/178/files#diff-948a751e87fecad4e01702996e31cfd86a12c675951bff5a7bcb94fef0af848a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>seed_hotel_images.py</strong><dd><code>Add management command for seeding hotel images</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/management/commands/seed_hotel_images.py

<li>Added new management command to seed hotel images.<br> <li> Generated random images for hotels using PIL.<br> <li> Ensured compatibility with existing hotel data.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/178/files#diff-e7ebceea0055b41eef521d4cd51afc94d7097540c4137996cc47e13e535c1595">+61/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: backend/pawtel/hotels/services.py [hunk 1]
```diff
@@ -197,14 +197,127 @@ def list_hotels(filters=None):
...
+        return serializer
+
+    def validate_upload_image(input_serializer, pk):
+        hotel = HotelService.retrieve_hotel(pk)
+        if not hotel or hotel.is_archived:
+            raise NotFound("Hotel does not exist or is archived.")
+        if not input_serializer.is_valid():
+            raise ValidationError(input_serializer.errors)
+        if hotel.images.count() >= 5:
+            raise ValidationError("A hotel cannot have more than 5 images.")
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:218]: Hace falta añadir manualmente {"detail": ...} en Validation error. En el de input_serializers.errors creo que no.

File: backend/pawtel/hotels/services.py [hunk 2]
```diff
@@ -197,14 +197,127 @@ def list_hotels(filters=None):
...
+        hotel = HotelService.retrieve_hotel(pk)
+        if not hotel or hotel.is_archived:
+            raise NotFound("Hotel does not exist or is archived.")
+
+        hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
+        if not hotel_image:
+            raise NotFound("Image not found or does not belong to the hotel.")
+
+    @staticmethod
+    def set_image_as_cover(request, pk, image_id):
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:310]: En las funciones que no necesiten request, quita el parámetro de request en la definición y no se lo pases.

File: backend/pawtel/hotels/services.py [hunk 3]
```diff
@@ -197,14 +197,127 @@ def list_hotels(filters=None):
...
+
+        hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
+        if not hotel_image:
+            raise NotFound("Image not found or does not belong to the hotel.")
+
+        if not input_serializer.is_valid():
+            raise ValidationError(input_serializer.errors)
+
+    @staticmethod
+    def update_image_to_hotel(input_serializer, pk, image_id):
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:266]: Igual que he dicho con el create, que creo que debería ser True. Y lo del current_cover_image.

File: backend/pawtel/hotels/services.py [hunk 4]
```diff
@@ -217,14 +217,17 @@ def validate_upload_image(input_serializer, pk):
...
             raise ValidationError("A hotel cannot have more than 5 images.")
 
-        if (
-            input_serializer.validated_data.get("is_cover")
-            and hotel.images.filter(is_cover=True).exists()
-        ):
-            raise ValidationError("Only one image can be set as the cover.")
-
     @staticmethod
     def upload_image_to_hotel(input_serializer):
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:221]: Ahí abajo no debería ser cuando is cover sea True? Cuando queremos que sea cover, entonces cogemos la cover actual y la cambiamos.

En vez de poner hotel_image_cover pon mejor current_cover_image, como has hecho en otros métodos, que está más claro.

Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:221]: Además:
- En el create, que cuente el número actual de imágenes del hotel. Si es 0, diga lo que diga el body, esta imagen va a ser cover sí o sí.
- En el update, que cuente el número actual de imágenes del hotel. Si es 1 y es la que estamos modificando, tiene que quedarse sí o sí como cover.

File: backend/pawtel/hotels/services.py [hunk 5]
```diff
@@ -259,21 +262,62 @@ def validate_update_image(input_serializer, pk, image_id):
...
         hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
         return input_serializer.update(hotel_image, input_serializer.validated_data)
 
     @staticmethod
     def delete_image_from_hotel(pk, image_id):
         hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
         hotel_image.delete()
+
+    @staticmethod
+    def rerieve_cover_image(request, pk):
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:285]: Ponle la t que se ha escapado (reTrieve) en vez de rerieve. Corrige los que le llamen.

File: backend/pawtel/hotels/views.py [hunk 6]
```diff
@@ -91,7 +94,7 @@ class HotelImageViewSet(viewsets.ViewSet):
     @action(
         detail=True,
         methods=["post"],
-        url_path="hotel_images/upload",
+        url_path="hotel-image/upload",
```
Comment (rafcasceb) [inline backend/pawtel/hotels/views.py:97]: No sé si te lo había dicho yo mal o qué, pero es hotel-images en plural, no en singular. Cámbialo en todos lados.

**Review by rafcasceb — CHANGES_REQUESTED** [review]

File: backend/pawtel/hotels/services.py [hunk 7]
```diff
@@ -197,14 +197,146 @@ def list_hotels(filters=None):
...
+
+    @staticmethod
+    def update_image_to_hotel(input_serializer, pk, image_id):
+        hotel = input_serializer.validated_data["hotel"]
+        current_image_count = hotel.images.count()
+
+        hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
+        if (
+            current_image_count == 1
+            and hotel_image == input_serializer.validated_data.get("image")
```
Comment (rafcasceb) [inline backend/pawtel/hotels/services.py:281]: Al final, con las validaciones que tenemos en los pasos previos ya no hace falta comprobar más que el current_image_count==1. Tan solo quitar la otra comprobación del if. Quedaría así, vamos:

```python
    @staticmethod
    def update_image_to_hotel(input_serializer, pk, image_id):
        hotel = input_serializer.validated_data["hotel"]
        current_image_count = hotel.images.count()
        hotel_image = HotelService.retrieve_image_from_hotel(pk, image_id)
        
        if current_image_count == 1:
            input_serializer.validated_data["is_cover"] = True
       
        ...
```

**Review by rafcasceb — CHANGES_REQUESTED** [review]

**Review by rafcasceb — APPROVED** [review]
Bien por fin.
